### PR TITLE
[Wiki Settings] Limiting Page Creation

### PR DESCRIPTION
**Pull request by ajainuary**
**Description**
This is a minor PR which adds a boolean `limitPageCreation` field to the `settings` object in the `wikiDoc`.
If this is set, the user has no option to create a new page.

**Resolved Issues**
This PR builds toward the Wiki Ownership feature.
Referenced Asana Task: https://app.asana.com/0/1121331595459324/1130445664537704

**How to test?**
Cannot be tested via Browser, need to manually make changes to the database.

### CODE REVIEW

File: frog/imports/client/Wiki/index.js [hunk 1]
```diff
@@ -74,7 +74,8 @@ type WikiCompStateT = {
   noInstance: ?boolean,
   username: string,
   isAnonymous: boolean,
-  settings?: WikiSettingsT
+  settings?: WikiSettingsT,
+  isOwner?: boolean
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js]: This should not be a definite type, it is already a truthy/falsy value

File: frog/imports/client/Wiki/index.js [hunk 2]
```diff
@@ -140,9 +141,10 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
 
     window.wikiDoc = this.wikiDoc;
     if (!this.props.embed) {
-      Mousetrap.bindGlobal('ctrl+n', () =>
-        this.setState({ createModalOpen: true })
-      );
+      Mousetrap.bindGlobal('ctrl+n', () => {
+        if (!this.state.settings?.noNewPage && !this.state.isOwner)
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js]: Is there an equivalent check in the server? We want to make sure the server also enforces this.

Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js]: `noNewPage` is not defined in `WikiSettingsT`

Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js]: 3\*2\*3 = 18 conditional possibilities, enforcing the presence of `settings` or making `isOwner` defined would help reduce this number

File: frog/imports/client/Wiki/index.js [hunk 3]
```diff
@@ -185,7 +187,8 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
     wikiStore.setPages(this.wikiDoc.data.pages);
     this.setState({
       pagesData: wikiStore.pages,
-      settings: this.wikiDoc.data.settings
+      settings: this.wikiDoc.data.settings,
+      isOwner: !!this.wikiDoc.data.owners.find(x => x === Meteor.userId())
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js]: NIT:
```suggestion
      isOwner: this.wikiDoc.data.owners.find(x => x === Meteor.userId()) === undefined
```

File: frog/imports/client/Wiki/index.js [hunk 4]
```diff
@@ -654,13 +657,15 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
...
         <h2>{this.wikiId}</h2>
         <ul>
-          <Button
-            variant="contained"
-            color="primary"
-            onClick={() => this.setState({ createModalOpen: true })}
-          >
-            + Create new page
-          </Button>
+          {(!this.state.settings?.limitPageCreation || this.state.isOwner) && (
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:662]: `limitPageCreation` is not defined in `WikiSettingsT`

Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:662]: What is the difference between `limitPageCreation` and `noNewPage`?

File: frog/imports/client/Wiki/index.js [hunk 5]
```diff
@@ -185,7 +187,8 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
     wikiStore.setPages(this.wikiDoc.data.pages);
     this.setState({
       pagesData: wikiStore.pages,
-      settings: this.wikiDoc.data.settings
+      settings: this.wikiDoc.data.settings,
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:192]: Make sure that `wikiDoc.data.settings` also has the type `WikiSettingsT` perhaps by moving the type definition to where `wikiDoc` is defined.

**Review by qgolsteyn — CHANGES_REQUESTED** [review]
Left some comments

File: frog/imports/client/Wiki/index.js (hunk 1 above)
Comment (ajainuary) [reply frog/imports/client/Wiki/index.js]: Made the respective changes

File: frog/imports/client/Wiki/index.js (hunk 2 above)
Comment (ajainuary) [reply frog/imports/client/Wiki/index.js]: Unfortunately, not yet. But we do have plans to implement that in the future

Comment (ajainuary) [reply frog/imports/client/Wiki/index.js]: I think this was an inadvertent error

Comment (ajainuary) [reply frog/imports/client/Wiki/index.js]: `isOwner` is now defined but cannot strictly enforce the presence of `settings` yet unless we can upgrade the existing wikis in production with the settings field in the database

File: frog/imports/client/Wiki/index.js (hunk 4 above)
Comment (ajainuary) [reply frog/imports/client/Wiki/index.js:662]: Made the respective changes

Comment (ajainuary) [reply frog/imports/client/Wiki/index.js:662]: I think this was an inadvertent error

File: frog/imports/client/Wiki/index.js (hunk 5 above)
Comment (ajainuary) [reply frog/imports/client/Wiki/index.js:192]: `wikiDoc` is the shareDB document, we will need to add validation rules in MongoDB

File: frog/imports/client/Wiki/index.js [hunk 6]
```diff
@@ -140,9 +143,10 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
 
     window.wikiDoc = this.wikiDoc;
     if (!this.props.embed) {
-      Mousetrap.bindGlobal('ctrl+n', () =>
-        this.setState({ createModalOpen: true })
-      );
+      Mousetrap.bindGlobal('ctrl+n', () => {
+        if (!this.state.settings?.limitPageCreation && !this.state.isOwner)
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:147]: So, we allow page creation if:
- `settings` is undefined (please consider giving settings a default value to simply the logic here), OR
- `limitPageCreation` is false, meaning page creation is allowed (consider changing limitPageCreation to allowPageCreation to remove the ! and improve readability, as per http://davej.io/2018/05/negative-naming.html), AND IN ANY CASE
- `isOwner` is false, **meaning that the owner of the wiki cannot add pages?** Make sure this is intended.

File: frog/imports/client/Wiki/index.js [hunk 7]
```diff
@@ -654,13 +659,15 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
...
         <h2>{this.wikiId}</h2>
         <ul>
-          <Button
-            variant="contained"
-            color="primary"
-            onClick={() => this.setState({ createModalOpen: true })}
-          >
-            + Create new page
-          </Button>
+          {(!this.state.settings?.limitPageCreation || this.state.isOwner) && (
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:662]: As the two conditionals are not the same, but seem to be affecting the same behaviour, make sure that this is intentional, and consider moving this logic to a separate function to keep the code DRY and avoid inconsistencies.

**Review by qgolsteyn — CHANGES_REQUESTED** [review]
More comments, yay

File: frog/imports/client/Wiki/index.js (hunk 2 above)
Comment (qgolsteyn) [reply frog/imports/client/Wiki/index.js]: We could add a default value for settings on the client side to at least clarify the default behaviour, would remove a number of `?` notations and improve the confidence in the code

File: frog/imports/client/Wiki/index.js [hunk 8]
```diff
@@ -185,7 +189,8 @@ class WikiComp extends React.Component<WikiCompPropsT, WikiCompStateT> {
     wikiStore.setPages(this.wikiDoc.data.pages);
     this.setState({
       pagesData: wikiStore.pages,
-      settings: this.wikiDoc.data.settings
+      settings: this.wikiDoc.data.settings,
+      isOwner: this.wikiDoc.data.owners.find(x => x === Meteor.userId()) === undefined
```
Comment (qgolsteyn) [inline frog/imports/client/Wiki/index.js:193]: Should it be `!== undefined`, meaning we found a match between the current user and one of the owners of the wiki?

**Review by qgolsteyn — COMMENTED** [review]
One last one

**Comment by houshuang** [comment]
@ajainuary 	Should we close this for your new combined PR?

**Comment by ajainuary** [comment]
Yes, this PR has been succeeded by #1784

**Pull request closed without merging**